### PR TITLE
[ci] Add mit-pdos/perennial

### PR DESCRIPTION
**Pull request by tchajed**
Add Perennial, a system for verifying concurrent, crash-safe systems. Perennial is built on top of Iris.

Perennial takes about 30 minutes to build, including all the verified examples (a replicated disk, a mail server, as well as some smaller ones).

**Kind:** infrastructure.

### CODE REVIEW

File: Makefile.ci [hunk 1]
```diff
@@ -35,6 +35,7 @@ CI_TARGETS= \
     ci-math-comp \
     ci-mtac2 \
     ci-paramcoq \
+		ci-perennial \
```
Comment (SkySkimmer) [inline Makefile.ci]: Weird indentation

**Comment by SkySkimmer** [comment]
>Perennial is built on top of Iris.

This will need a bit more work then.
- the job in gitlab yml needs to depend on the iris job
- the dependency needs to be added in Makefile.ci

Perhaps we should separate lambdarust from the main iris job to allow parallelism

Alternatively we could build after lambdarust in the iris job itself.

**Comment by tchajed** [comment]
Oops, forgot to mention: we build on top of Iris, but we include it as a git submodule and build it from our own Makefile. (@Zimmi48 has said we shouldn't do this, but it works extremely reliably and lets us easily pin Iris to a particular commit.) It adds about 5 CPU-minutes to our build.

**Comment by tchajed** [comment]
Just for the record, I finally decided to do this in response to #10476, which broke Perennial in a bunch of subtle ways.

**Comment by ejgallego** [comment]
@tchajed good point, imo the real problem here is lack of a known problems file, for example it is known that export has a few quirks so indeed it is not very safe to use if you want stability.

**Comment by maximedenes** [comment]
> Just for the record, I finally decided to do this in response to #10476, which broke Perennial in a bunch of subtle ways.

I'm not sure what to think about this. #10476 is a bug fix, and it took me ~2 months to fix all developments (in our CI) which were relying on the bug. If we keep adding developments to the CI, aren't we going to reach a point where it will be infeasible to fix bugs?

Note that I'm not opposing to the addition of this development to the CI per se, but the motivation should be that it exercises parts of Coq that are not already covered by our current CI IMHO. Is it the case?

> for example it is known that export has a few quirks so indeed it is not very safe to use if you want stability.

It wasn't known by me before I noticed this bug. And since it is widely used in stdlib, I'm pretty sure a number of users were unaware of the problem(s).

**Comment by maximedenes** [comment]
> it took me ~2 months to fix all developments (in our CI) which were relying on the bug

(mainly because of the number of things to build, inspect and fix, and because of the synchronization costs, not because any kind of cleverness was required)

**Comment by Zimmi48** [comment]
> I'm not sure what to think about this. #10476 is a bug fix, and it took me ~2 months to fix all developments (in our CI) which were relying on the bug. If we keep adding developments to the CI, aren't we going to reach a point where it will be infeasible to fix bugs?

In some cases, we should request users to fix their projects on their own. Give them advanced notice (e.g. two months), and clear instructions on how to test their projects with the new code.

However, @tchajed says "a bunch of subtle ways". So, if the problem is not easy to identify, it might deserve so more thinking. If many projects had to be fixed, maybe the warning that this is a compatibility breaking change in the changelog should be written in stronger terms. And can we think of some way to make the errors clearer? Or can we put a deprecated compatibility option that gets activated in compat mode?

> Oops, forgot to mention: we build on top of Iris, but we include it as a git submodule and build it from our own Makefile. (@Zimmi48 has said we shouldn't do this, but it works extremely reliably and lets us easily pin Iris to a particular commit.) It adds about 5 CPU-minutes to our build.

I wouldn't block this PR on this issue, especially if it's just about 5 minutes of build time, but at some point in the future we will *have to* rationalize dependency management in CI, and impose a unique model (I don't know yet what it will look like). At which point, we may end up expelling projects from CI if no common ground can be found with the maintainers.

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -733,6 +733,9 @@ plugin:ci-mtac2:
 plugin:ci-paramcoq:
   extends: .ci-template
 
+plugin:ci-perennial:
+  extends: .ci-template
```
Comment (Zimmi48) [inline .gitlab-ci.yml]: You might want to try the flambda template to see if it's faster.

**Review by Zimmi48 — APPROVED** [review]

**Comment by tchajed** [comment]
@maximedenes I don't want to add to your burden in fixing downstream projects when there are incompatible changes - what I want is for Coq developers to be aware when a change causes non-trivial breakage, since in that case somebody will have to fix all the developments that we would like to continue to work. In this case it seems like that was quite clear without Perennial. I would also love to be CC'd on issues that break many projects, Iris, LambdaRust, or Perennial specifically.

I fully agree with @Zimmi48 on the point that if this was so difficult, perhaps it should require some thought on the best way to deploy it. In this case I was stymied by a lack of documentation on what I should actually do. From my perspective, there was no bug; the behavior just changed.

**Comment by tchajed** [comment]
As far as being different from other projects - it's hard to say, but I think we use Iris in some pretty odd ways that are different from LambdaRust. That said, having one Iris project is much better than none.

We're actively working on Perennial and I only use Coq master for daily development, so if you tell me it's going to break on a Coq PR, I'l switch to that PR branch and fix it. If we stop actively working on it and it becomes a burden, feel free to drop it from CI.

**Comment by maximedenes** [comment]
> I don't want to add to your burden in fixing downstream projects when there are incompatible changes - what I want is for Coq developers to be aware when a change causes non-trivial breakage, since in that case somebody will have to fix all the developments that we would like to continue to work. In this case it seems like that was quite clear without Perennial. I would also love to be CC'd on issues that break many projects, Iris, LambdaRust, or Perennial specifically.

I agree, CI plays a key role in identifying such changes, and I like your idea of improving the communication around breaking changes.

I didn't want to sound too alarmist, as I still hope this bug was a bit extreme, in that the symptoms were often very distant from the observed change. The vernacular language has a lot of implicit relationships between objects (tactics and hints in scope, type inference and coercions in scope,...) and this fix was precisely about the scoping of containers.

> maybe the warning that this is a compatibility breaking change in the changelog should be written in stronger terms

Yep, I asked for suggestions in the PR thread. Not too late to improve the phrasing, though.

> And can we think of some way to make the errors clearer?

I thought about it, but didn't know where to start. How do you detect that a user wrongly expects a hint to be in scope? Or a notation scope to be open?

> Or can we put a deprecated compatibility option that gets activated in compat mode?

I am going to somewhat contradict what I said earlier. I would still prefer to put all developments in the CI and have to fix them all, rather than add more compat options. The first is essentially linear complexity (porting twice as many developments takes twice as much time), whereas testing and maintaining all combinations of options is exponential.

> In this case I was stymied by a lack of documentation on what I should actually do. From my perspective, there was no bug; the behavior just changed.

I can easily understand the feeling. In fact the easiest time to see that something was wrong was at the time the code was introduced, so hopefully it would be caught by our review system today. Then, for the low-level libraries which use `Export`, the behavior was sometimes obviously wrong, but AFAIK it was not reported. At the level of stdpp you could still figure it out, but it became more subtle. At the level of Iris, I needed to interact with the Iris developers (who kindly helped me) to understand what their intent was. I'm guessing that at the level of your library, it is probably really hard to understand what changed under your feet, and how it relates to the fix.

So to summarize what I wanted to say: I do expect to spend a significant amount of time interacting with our users when such changes or fixes are integrated, and I actually like it. What I'm wondering is whether the CI model is the right tool to synchronize all Coq developments, or if we want two things, a CI to make sure we are aware of breaking changes before they happen, and something else to work at a larger scale on maintaining the whole Coq Universe consistent.

**Comment by ejgallego** [comment]
> It wasn't known by me before I noticed this bug. And since it is widely used in stdlib, I'm pretty sure a number of users were unaware of the problem(s).

Indeed not about this specific bug, but as commented [here](https://github.com/coq/coq/issues/10474#issuecomment-508121042) a few developers (including myself) are wary of `Export`, mainly due to discussions in WGs.

> mainly because of the number of things to build, inspect and fix, and because of the synchronization costs, not because any kind of cleverness was required

The good news on that is that the "overlay" toolchain should greatly improve in the next few months, to the point you would expect everything but the actual code fixes being done by tools.

**Comment by Zimmi48** [comment]
>whereas testing and maintaining all combinations of options is exponential

I was only suggesting a temporary option to help with porting. For instance, the changelog entry could say: if you don't understand why your project breaks with this update, you might want to try turning this option on to see if it is because of this change. But the option would be deprecated from the start.

I am only trying to think of what is the best porting scenario (without having participated in the PR discussion, indeed).

**Comment by ejgallego** [comment]
> I tried using flambda and it was a bit slower ([normal run](https://gitlab.com/coq/coq/-/jobs/299421977) vs [flamba](https://gitlab.com/coq/coq/-/jobs/299462171)), so I reverted that.

Numbers obtained this way are highly unreliable tho, workers are just not able to provide any kind of reliable timing.

**Comment by tchajed** [comment]
> Numbers obtained this way are highly unreliable tho, workers are just not able to provide any kind of reliable timing.

I agree; if it were significantly faster from just one run I would have used flambda, but otherwise I'd rather go with the standard setup.

**Comment by ejgallego** [comment]
At least for OCaml 4.07, we have been seen very nice speedups for users of flambda-optimized Coq, I would be surprised if in your setup you don't see any gain, I would indeed try in a more controlled environment [don't forget to pass the right configure flags to Coq as detailed in the INSTALL file]

**Comment by gares** [comment]
> In some cases, we should request users to fix their projects on their own. Give them advanced notice (e.g. two months), and clear instructions on how to test their projects with the new code.

I could not agree more... But some may fail to miss the deadline, and multiple deadlines could be raised at the same time, making synchronization hard. It is not as easy as it seems.

To me this option is only feasible if we manage to also get a real test suite: today ci helps active projects to be ported *and* the detection of breakage. The former could go at a lower pace, but as of today, a lower pace there means we can't detect regressions until these projects are fixed... and that can't possibly work.

Maybe it's time to start writing real tests whenever we add/update/fix code?

**Comment by Zimmi48** [comment]
>Maybe it's time to start writing real tests whenever we add/update/fix code?

Could you be more specific on what you mean by this?

FTR, I also think that it is time to improve our test-suite based on our experience with external projects in CI. My plan to do that was to list all the builds which resulted in a failure in an external project but no failure in the test-suite (already done, and more than half of the failures in external projects are currently uncaught by the test-suite), then to try to extract a minimal reproducible example out of each of these failures that could be added to the test-suite.

---

@coq/ci-maintainers it seems that the PR is approved and ready to be merged.

**Comment by ejgallego** [comment]
Regarding testing we already have coverage data so that should provide a good lead to improve the test suite.

Regarding this PR, it is not so clear to me that we should not use the flambda template, @SkySkimmer ?

**Comment by SkySkimmer** [comment]
Why do we not use flambda for everything again?

**Comment by tchajed** [comment]
Ok, I switched to flambda.

**Comment by ejgallego** [comment]
> Why do we not use flambda for everything again?

To still test the regular toolchain a bit.

**Comment by gares** [comment]
> Could you be more specific on what you mean by this?

Well, it is customary to write functional tests for your code. We have a lot of code, but no functional tests. A way to get some could be to apply a rule like: "you touch/add X, you add some tests for X".

**Review by ejgallego — APPROVED** [review]
Looks good to me, any further comment?

**Comment by JasonGross** [comment]
> FTR, I also think that it is time to improve our test-suite based on our experience with external projects in CI. My plan to do that was to list all the builds which resulted in a failure in an external project but no failure in the test-suite (already done, and more than half of the failures in external projects are currently uncaught by the test-suite), then to try to extract a minimal reproducible example out of each of these failures that could be added to the test-suite.

@Zimmi48 this process can be mostly automated (and I can help with that).  My bug minimizer can extract stand-alone somewhat-minimal examples of code that fail with a particular error message in one version of Coq and succeed with another version.  (I think it currently only works when it inlines all dependencies first, but I can update it so you can give it two different working directories for the project each built with a different version of Coq.)  So if you automate the process of "set up the environment so the good and bad versions of Coq are in particular places and the protect is checked out to the right commit and built with each version of Coq up to the failing file and you have the path to the failing file somewhere", I can automate the process of turning that into a smaller stand-alone example.

**Comment by Zimmi48** [comment]
Yes, I was indeed thinking of using your bug minimizer, and I thank you for your offer to add what's needed. I will follow up on this. (Currently in the last week before submitting my thesis, and afterwards I need to take a break *and* I have promised to a lot of people that I would do a lot of things, but I will definitely follow up on this!)
